### PR TITLE
FEAT-#2087 Added benchmarks test suite

### LINKED ISSUE #2087: Add micro-benchmarks performance test suite to Modin
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. What kind of performance improvements would you like to see with this new API?

We need a place to easily add micro-benchmarks for individual functions, possibly compare them with Modin. I think it should be based on `pytest-benchmark`.

### PR DESCRIPTION
**Pull request by gshimansky**
Benchmarks added for the following functions:
- DataFrame.sum()
- DataFrame.agg(sum)
- DataFrame.agg("sum")
- GroupBy.sum()
- GroupBy.agg(sum)

Signed-off-by: Gregory Shimansky <gregory.shimansky@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2087 
- [x] tests added and passing

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -12,6 +12,7 @@ tables
 scipy
 s3fs>=0.4.2
 pytest
+pytest-benchmark
```
Comment (anmyachev) [inline requirements.txt:15]: This also needed for `environment.yml`.

File: modin/pandas/test/test_benchmarks.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
+# Licensed to Modin Development Team under one or more contributor license agreements.
```
Comment (anmyachev) [inline modin/pandas/test/test_benchmarks.py:1]: What about `ci/benchmarks` as a place for all performance test?

Comment (gshimansky) [reply modin/pandas/test/test_benchmarks.py:1]: > What about `ci/benchmarks` as a place for all performance test?

I don't think they are used anywhere.

File: requirements.txt (hunk 1 above)
Comment (gshimansky) [reply requirements.txt:15]: Ok

File: modin/pandas/test/test_benchmarks.py (hunk 2 above)
Comment (devin-petersohn) [reply modin/pandas/test/test_benchmarks.py:1]: It might be nice to separate benchmarks from unit tests.

**Review by devin-petersohn — COMMENTED** [review]
Very nice! Where are we planning to run these?

File: modin/pandas/test/test_benchmarks.py (hunk 2 above)
Comment (gshimansky) [reply modin/pandas/test/test_benchmarks.py:1]: What is your suggestion? Move this file into `benchmarks` subfolder?

Comment (devin-petersohn) [reply modin/pandas/test/test_benchmarks.py:1]: Yes, I believe that is also @anmyachev's suggestion as well.

Comment (gshimansky) [reply modin/pandas/test/test_benchmarks.py]: Ok I don't insist on this location and moved this file to `ci/benchmarks`.
Other files in that place however use their own test framework and aren't pytest based.

File: ci/benchmarks/test_benchmarks.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+dataframes_cache = dict()
+
+
+def gen_int_data(ncols, nrows, rand_low, rand_high):
+    cache_key = ("int", ncols, nrows, rand_low, rand_high)
+    if cache_key in data_cache:
+        return data_cache[cache_key]
+
+    logging.info(
+        "Generating int data {}x{} [{}-{}]".format(ncols, nrows, rand_low, rand_high)
```
Comment (YarShev) [inline ci/benchmarks/test_benchmarks.py:49]: I think as the first dimension should be `nrows` and as the second one should be `ncols`.

File: ci/benchmarks/test_benchmarks.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    return data
+
+
+def gen_str_int_data(ncols, nrows, rand_low, rand_high):
+    cache_key = ("str_int", ncols, nrows, rand_low, rand_high)
+    if cache_key in data_cache:
+        return data_cache[cache_key]
+
+    logging.info(
+        "Generating str_int data {}x{} [{}-{}]".format(
```
Comment (YarShev) [inline ci/benchmarks/test_benchmarks.py:65]: The same comment about dimensions.

File: ci/benchmarks/test_benchmarks.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        assert False
+
+
+def generate_dataframe(impl, data_type, ncols, nrows, rand_low, rand_high):
+    cache_key = (impl, data_type, ncols, nrows, rand_low, rand_high)
+    if cache_key in dataframes_cache:
+        return dataframes_cache[cache_key]
+
+    logging.info(
+        "Allocating {} DataFrame {}: {}x{} [{}-{}]".format(
```
Comment (YarShev) [inline ci/benchmarks/test_benchmarks.py:92]: The same comment about dimensions.

File: ci/benchmarks/test_benchmarks.py (hunk 3 above)
Comment (gshimansky) [reply ci/benchmarks/test_benchmarks.py:49]: Why? Cols*rows is X*Y dimensions.

Comment (YarShev) [reply ci/benchmarks/test_benchmarks.py:49]: Yes, but when we call `df.shape`, we see row count as the first number and column count as the second number. An example:
```python
import modin.pandas as pd
df = pd.DataFrame([[1,1,1],[1,1,1]])
df
   0  1  2
0  1  1  1
1  1  1  1
df.shape
(2, 3)
```
That's why I think, when we write something like this "data ...x...", it's better to use row count as the first number.

Comment (gshimansky) [reply ci/benchmarks/test_benchmarks.py]: Ok I changed log output to remove any ambiguity

**Comment by YarShev** [comment]
@gshimansky , are there any thoughts on expanding of the benchmarks?

**Comment by gshimansky** [comment]
> @gshimansky , are there any thoughts on expanding of the benchmarks?

Sure, this commit is just the start. Right now I have plans for `merge`, `mul`, `mask based indexing`, maybe more will emerge while benchmarks performance is investigated.

**Comment by YarShev** [comment]
Okay, I am just interesting on how will benchmarks architecture be organized. There changes looks good to me. @devin-petersohn , @anmyachev , any comments?

**Comment by gshimansky** [comment]
> Okay, I am just interesting on how will benchmarks architecture be organized. These changes look good to me. @devin-petersohn , @anmyachev , any comments?

Currently I plan to add them to the same file. It is still very small. It is grows big, we can split it based on whatever criteria that we like, but this moment is quite far from now.

**Comment by YarShev** [comment]
@gshimansky , please, answer this [question](https://github.com/modin-project/modin/pull/2103#pullrequestreview-491753641).

**Comment by gshimansky** [comment]
> @gshimansky , please, answer this [question](https://github.com/modin-project/modin/pull/2103#pullrequestreview-491753641).

If you mean regular CI runs, I don't have plans for benchmarks so far. If you mean irregular runs to test performance, you can try to run these benchmarks anywhere you run other Modin applications.

**Comment by YarShev** [comment]
As I see, we will track performance manually for now. It would be good to think of automatic checking the performance in future.

**Review by devin-petersohn — APPROVED** [review]
LGTM, thanks @gshimansky!

**Comment by devin-petersohn** [comment]
I will let @anmyachev take a final pass to verify prior comments are resolved.

**Review by anmyachev — APPROVED** [review]

**Pull request merged**